sky/utils/controller_utils.py

```diff
@@ -300,7 +319,34 @@ def get_controller_resources(
                     err=f'Expected exactly one resource, got '
                     f'{len(controller_resources)} resources: '
                     f'{controller_resources}'))
-    return list(controller_resources)[0]
+    controller_resources_to_use: resources.Resources = list(
+        controller_resources)[0]
+
+    controller_exist = (global_user_state.get_cluster_from_name(controller_name)
+                        is not None)
+    if controller_exist or controller_resources_to_use.cloud is not None:
+        return {controller_resources_to_use}
+
+    # If the controller and replicas are from the same cloud, it should
+    # provide better connectivity. We will let the controller choose from
+    # the clouds of the resources if the controller does not exist.
+    # TODO(tian): Consider respecting the regions/zones specified for the
+    # resources as well.
+    requested_clouds: Set['clouds.Cloud'] = set()
+    for res in task_resources:
+        # cloud is an object and will not be able to be distinguished by set.
+        # Here we manually check if the cloud is in the set.
+        if res.cloud is not None and not clouds.cloud_in_iterable(
+                res.cloud, requested_clouds):
+            if not clouds.cloud_in_iterable(res.cloud,
+                                            _CONTROLLER_CLOUD_BLACKLIST):
```

Comment:
In #3377, we adds the `HOST_CONTROLLER` implementation feature in the `Cloud` to filter out the clouds that do not suitable for hosting controllers. We might want to do similar thing, instead of having another list for the black list:

https://github.com/skypilot-org/skypilot/blob/87330bd828ae9b17b656a67239f74b889fc48fb5/sky/clouds/cloud.py#L44
https://github.com/skypilot-org/skypilot/blob/87330bd828ae9b17b656a67239f74b889fc48fb5/sky/clouds/kubernetes.py#L105-L106

Reply:
Sounds good! Seems like #3377 is still under reviewing. Should we extract those part into a separate PR and merge it first? cc @romilbhardwaj for a look here

Reply:
sounds good @cblmemo - I'll create a separate PR for `CloudImplementationFeatures.HOST_CONTROLLERS`